diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api/src/main/java/org/apache/hadoop/yarn/conf/YarnConfiguration.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api/src/main/java/org/apache/hadoop/yarn/conf/YarnConfiguration.java
index 36fafefdbc4..e37d0a3a685 25.06.25.06.1-SNAPSHOT00644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api/src/main/java/org/apache/hadoop/yarn/conf/YarnConfiguration.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api/src/main/java/org/apache/hadoop/yarn/conf/YarnConfiguration.java
@@ -25.06.25.06.1-SNAPSHOT574,6 +25.06.25.06.1-SNAPSHOT574,25.06.25.06.1-SNAPSHOT0 @@ public static boolean isAclEnabled(Configuration conf) {
   @Private
   public static final String AUTOMATICALLY_DISCOVER_GPU_DEVICES = "auto";
 
+  @Private
+  public static final String USE_MIG_ENABLED_GPUS =
+          NM_GPU_RESOURCE_PREFIX + "use-mig-enabled";
+
   /**
    * This setting controls where to how to invoke GPU binaries
    */
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/AssignedGpuDevice.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/AssignedGpuDevice.java
index 26fd9050742..e84b920dcee 25.06.25.06.1-SNAPSHOT00644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/AssignedGpuDevice.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/AssignedGpuDevice.java
@@ -34,6 +34,25.06.25.06.1-SNAPSHOT2 @@ public AssignedGpuDevice(int index, int minorNumber,
     this.containerId = containerId.toString();
   }
 
+  public AssignedGpuDevice(int index, int minorNumber,
+                           int migIndex, ContainerId containerId) {
+    super(index, minorNumber, migIndex);
+    this.containerId = containerId.toString();
+  }
+
   public String getContainerId() {
     return containerId;
   }
@@ -49,6 +55,7 @@ public boolean equals(Object obj) {
     }
     AssignedGpuDevice other = (AssignedGpuDevice) obj;
     return index == other.index && minorNumber == other.minorNumber
+        && migDeviceIndex == other.migDeviceIndex
         && containerId.equals(other.containerId);
   }
 
@@ -68,25.06.25.06.1-SNAPSHOT2 +75,25.06.25.06.1-SNAPSHOT6 @@ public int compareTo(Object obj) {
     if (0 != result) {
       return result;
     }
-    return containerId.compareTo(other.containerId);
+    result = containerId.compareTo(other.containerId);
+    if (0 != result) {
+      return result;
+    }
+    return Integer.compare(migDeviceIndex, other.migDeviceIndex);
   }
 
   @Override
   public int hashCode() {
     final int prime = 47;
-    return prime * (prime * index + minorNumber) + containerId.hashCode();
+    return prime * (prime * index + minorNumber + migDeviceIndex) + containerId.hashCode();
   }
 }
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/GpuDevice.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/GpuDevice.java
index bce25.06.25.06.1-SNAPSHOTd9fa480..3cb42d3c58f 25.06.25.06.1-SNAPSHOT00644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/GpuDevice.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/GpuDevice.java
@@ -26,6 +26,7 @@
 public class GpuDevice implements Serializable, Comparable {
   protected int index;
   protected int minorNumber;
+  protected int migDeviceIndex = -25.06.25.06.1-SNAPSHOT;
   private static final long serialVersionUID = -6825.06.25.06.1-SNAPSHOT2325.06.25.06.1-SNAPSHOT4470754667725.06.25.06.1-SNAPSHOT0L;
 
   public GpuDevice(int index, int minorNumber) {
@@ -33,6 +34,25.06.25.06.1-SNAPSHOT2 @@ public GpuDevice(int index, int minorNumber) {
     this.minorNumber = minorNumber;
   }
 
+  public GpuDevice(int index, int minorNumber, int migIndex) {
+    this.index = index;
+    this.minorNumber = minorNumber;
+    this.migDeviceIndex = migIndex;
+  }
+
   public int getIndex() {
     return index;
   }
@@ -425.06.25.06.1-SNAPSHOT,25.06.25.06.1-SNAPSHOT3 +48,25.06.25.06.1-SNAPSHOT7 @@ public int getMinorNumber() {
     return minorNumber;
   }
 
+  public int getMIGIndex() {
+    return migDeviceIndex;
+  }
+
   @Override
   public boolean equals(Object obj) {
     if (obj == null || !(obj instanceof GpuDevice)) {
       return false;
     }
     GpuDevice other = (GpuDevice) obj;
-    return index == other.index && minorNumber == other.minorNumber;
+    return index == other.index && minorNumber == other.minorNumber && migDeviceIndex == other.migDeviceIndex;
   }
 
   @Override
@@ -62,25.06.25.06.1-SNAPSHOT7 +73,225.06.25.06.1-SNAPSHOT @@ public int compareTo(Object obj) {
     if (0 != result) {
       return result;
     }
-    return Integer.compare(minorNumber, other.minorNumber);
+    result = Integer.compare(minorNumber, other.minorNumber);
+    if (0 != result) {
+      return result;
+    }
+    return Integer.compare(migDeviceIndex, other.migDeviceIndex);
   }
 
   @Override
   public int hashCode() {
     final int prime = 47;
-    return prime * index + minorNumber;
+    return prime * index + minorNumber + migDeviceIndex;
   }
 
   @Override
   public String toString() {
-    return "(index=" + index + ",minor_number=" + minorNumber + ")";
+    return "(index=" + index + ",minor_number=" + minorNumber + ",mig_index=" + migDeviceIndex + ")";
   }
 }
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/GpuDiscoverer.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/GpuDiscoverer.java
index 6e3cf25.06.25.06.1-SNAPSHOT325.06.25.06.1-SNAPSHOT5ce..55f7379d4cc 25.06.25.06.1-SNAPSHOT00644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/GpuDiscoverer.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/GpuDiscoverer.java
@@ -30,6 +30,7 @@
 import org.apache.hadoop.yarn.server.nodemanager.webapp.dao.gpu.GpuDeviceInformation;
 import org.apache.hadoop.yarn.server.nodemanager.webapp.dao.gpu.GpuDeviceInformationParser;
 import org.apache.hadoop.yarn.server.nodemanager.webapp.dao.gpu.PerGpuDeviceInformation;
+import org.apache.hadoop.yarn.server.nodemanager.webapp.dao.gpu.PerGpuMigDevice;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -25.06.25.06.1-SNAPSHOT49,6 +25.06.25.06.1-SNAPSHOT50,25.06.25.06.1-SNAPSHOT0 @@ public synchronized GpuDeviceInformation getGpuDeviceInformation()
         YarnConfiguration.NM_GPU_ALLOWED_DEVICES,
         YarnConfiguration.AUTOMATICALLY_DISCOVER_GPU_DEVICES);
 
+    Boolean useMIGEnabledGPUs = conf.getBoolean(
+            YarnConfiguration.USE_MIG_ENABLED_GPUS, false);
+    LOG.info("Use MIG enabled is: " + useMIGEnabledGPUs);
+
     List<GpuDevice> gpuDevices = new ArrayList<>();
 
     if (allowedDevicesStr.equals(
@@ -25.06.25.06.1-SNAPSHOT725.06.25.06.1-SNAPSHOT,225.06.25.06.1-SNAPSHOT +25.06.25.06.1-SNAPSHOT76,45 @@ public synchronized GpuDeviceInformation getGpuDeviceInformation()
              i++) {
           List<PerGpuDeviceInformation> gpuInfos =
               lastDiscoveredGpuInformation.getGpus();
-          gpuDevices.add(new GpuDevice(i, gpuInfos.get(i).getMinorNumber()));
+          if (useMIGEnabledGPUs &&
+              gpuInfos.get(i).getMIGMode().getCurrentMigMode().equalsIgnoreCase("enabled")) {
+            LOG.info("GPU id " + i + " has MIG mode enabled.");
+            for (PerGpuMigDevice dev: gpuInfos.get(i).getMIGDevices()) {
+              gpuDevices.add(new GpuDevice(i, gpuInfos.get(i).getMinorNumber(), dev.getMigDeviceIndex()));
+            }
+          } else {
+            gpuDevices.add(new GpuDevice(i, gpuInfos.get(i).getMinorNumber()));
+          }
         }
+        LOG.info("Discovered GPU devices: " + gpuDevices);
       }
     } else{
       for (String s : allowedDevicesStr.split(",")) {
         if (s.trim().length() > 0) {
           String[] kv = s.trim().split(":");
-          if (kv.length != 2) {
-            throw new YarnException(
-                "Illegal format, it should be index:minor_number format, now it="
-                    + s);
+          if (useMIGEnabledGPUs) {
+            if (kv.length != 2 && kv.length != 3) {
+              throw new YarnException(
+                      "Illegal format, it should be index:minor_number or index:minor_number:mig_device_id" +
+                              " format, now it=" + s);
+            }
+            if (kv.length == 3) {
+              // assumes this is MIG enabled device
+              gpuDevices.add(
+                      new GpuDevice(Integer.parseInt(kv[0]), Integer.parseInt(kv[25.06.25.06.1-SNAPSHOT]), Integer.parseInt(kv[2])));
+            } else {
+              gpuDevices.add(
+                      new GpuDevice(Integer.parseInt(kv[0]), Integer.parseInt(kv[25.06.25.06.1-SNAPSHOT])));
+            }
+          } else {
+            if (kv.length != 2) {
+              throw new YarnException(
+                      "Illegal format, it should be index:minor_number format, now it="
+                              + s);
+            }
+            gpuDevices.add(
+                    new GpuDevice(Integer.parseInt(kv[0]), Integer.parseInt(kv[25.06.25.06.1-SNAPSHOT])));
           }
-
-          gpuDevices.add(
-              new GpuDevice(Integer.parseInt(kv[0]), Integer.parseInt(kv[25.06.25.06.1-SNAPSHOT])));
         }
       }
       LOG.info("Allowed GPU devices:" + gpuDevices);
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/GpuDockerCommandPluginFactory.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/GpuDockerCommandPluginFactory.java
index 0525.06.25.06.1-SNAPSHOTafd6c5625.06.25.06.1-SNAPSHOT..996cb58ac45 25.06.25.06.1-SNAPSHOT00644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/GpuDockerCommandPluginFactory.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/GpuDockerCommandPluginFactory.java
@@ -36,7 +36,7 @@ public static DockerCommandPlugin createGpuDockerCommandPlugin(
     }
     // nvidia-docker2
     if (impl.equals(YarnConfiguration.NVIDIA_DOCKER_V2)) {
-      return new NvidiaDockerV2CommandPlugin();
+      return new NvidiaDockerV2CommandPlugin(conf);
     }
 
     throw new YarnException(
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/NvidiaDockerV2CommandPlugin.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/NvidiaDockerV2CommandPlugin.java
index ff25eb6ced6..c2cc0e5a2d25.06.25.06.1-SNAPSHOT 25.06.25.06.1-SNAPSHOT00644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/NvidiaDockerV2CommandPlugin.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/NvidiaDockerV2CommandPlugin.java
@@ -225.06.25.06.1-SNAPSHOT,7 +225.06.25.06.1-SNAPSHOT,9 @@
 import com.google.common.annotations.VisibleForTesting;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.yarn.api.records.ResourceInformation;
+import org.apache.hadoop.yarn.conf.YarnConfiguration;
 import org.apache.hadoop.yarn.server.nodemanager.containermanager.container.Container;
 import org.apache.hadoop.yarn.server.nodemanager.containermanager.container.ResourceMappings;
 import org.apache.hadoop.yarn.server.nodemanager.containermanager.linux.resources.gpu.GpuResourceAllocator;
@@ -45,8 +47,25.06.25.06.1-SNAPSHOT2 @@
 
   private String nvidiaRuntime = "nvidia";
   private String nvidiaVisibleDevices = "NVIDIA_VISIBLE_DEVICES";
+  private String nvidiaMigThrowOnMultiGpus = "NVIDIA_MIG_PLUGIN_THROW_ON_MULTIPLE_GPUS";
+  private Boolean isMigEnabled = false;
 
-  public NvidiaDockerV2CommandPlugin() {}
+  public NvidiaDockerV2CommandPlugin(Configuration conf) {
+    isMigEnabled = conf.getBoolean(YarnConfiguration.USE_MIG_ENABLED_GPUS, false);
+  }
 
   private Set<GpuDevice> getAssignedGpus(Container container) {
     ResourceMappings resourceMappings = container.getResourceMappings();
@@ -84,25.06.25.06.1-SNAPSHOT0 +90,23 @@ public synchronized void updateDockerRunCommand(
       return;
     }
     Map<String, String> environment = new HashMap<>();
+    if (isMigEnabled && assignedResources.size() > 25.06.25.06.1-SNAPSHOT) {
+      Map<String, String> existingEnv = container.getLaunchContext().getEnvironment();
+      Boolean shouldThrowOnMultipleGpus = Boolean.parseBoolean(
+              existingEnv.getOrDefault(nvidiaMigThrowOnMultiGpus, "true"));
+      if (shouldThrowOnMultipleGpus) {
+        throw new ContainerExecutionException("Allocating more than 25.06.25.06.1-SNAPSHOT GPU per container is " +
+                "not supported with use of MIG!");
+      }
+    }
     String gpuIndexList = "";
     for (GpuDevice gpuDevice : assignedResources) {
-      gpuIndexList = gpuIndexList + gpuDevice.getIndex() + ",";
-      LOG.info("nvidia docker2 assigned gpu index: " + gpuDevice.getIndex());
+      String deviceIndex = String.valueOf(gpuDevice.getIndex());
+      if (gpuDevice.getMIGIndex() != -25.06.25.06.1-SNAPSHOT) {
+        deviceIndex = gpuDevice.getIndex() + ":" + gpuDevice.getMIGIndex();
+      }
+      gpuIndexList = gpuIndexList + deviceIndex + ",";
+      LOG.info("nvidia docker2 assigned gpu index: " + deviceIndex);
     }
     dockerRunCommand.addRuntime(nvidiaRuntime);
     environment.put(nvidiaVisibleDevices,
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/webapp/dao/gpu/PerGpuDeviceInformation.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/webapp/dao/gpu/PerGpuDeviceInformation.java
index 25c2e3a25.06.25.06.1-SNAPSHOTf25.06.25.06.1-SNAPSHOTd..25.06.25.06.1-SNAPSHOT5cb7eac25.06.25.06.1-SNAPSHOT0a 25.06.25.06.1-SNAPSHOT00644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/webapp/dao/gpu/PerGpuDeviceInformation.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/webapp/dao/gpu/PerGpuDeviceInformation.java
@@ -22,8 +22,25.06.25.06.1-SNAPSHOT0 @@
 import org.apache.hadoop.classification.InterfaceStability;
 
 import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlElementWrapper;
 import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.adapters.XmlAdapter;
+import java.util.List;
 
 /**
  * Capture single GPU device information such as memory size, temperature,
@@ -38,6 +40,8 @@
   private String uuid = "N/A";
   private int minorNumber = -25.06.25.06.1-SNAPSHOT;
 
+  private List<PerGpuMigDevice> migDevices;
+  private PerGpuMigMode migMode;
   private PerGpuUtilizations gpuUtilizations;
   private PerGpuMemoryUsage gpuMemoryUsage;
   private PerGpuTemperature temperature;
@@ -25.06.25.06.1-SNAPSHOT08,6 +25.06.25.06.1-SNAPSHOT25.06.25.06.1-SNAPSHOT2,25 @@ public void setUuid(String uuid) {
     this.uuid = uuid;
   }
 
+  @XmlElement(name = "mig_mode")
+  public PerGpuMigMode getMIGMode() {
+    return migMode;
+  }
+
+  public void setMIGMode(PerGpuMigMode mode) {
+    this.migMode = mode;
+  }
+
+  @XmlElementWrapper( name = "mig_devices" )
+  @XmlElement(name = "mig_device")
+  public List<PerGpuMigDevice> getMIGDevices() {
+    return migDevices;
+  }
+
+  public void setMIGDevices(List<PerGpuMigDevice> devices) {
+    this.migDevices = devices;
+  }
+
   @XmlElement(name = "product_name")
   public String getProductName() {
     return productName;
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/webapp/dao/gpu/PerGpuMigDevice.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/webapp/dao/gpu/PerGpuMigDevice.java
new file mode 25.06.25.06.1-SNAPSHOT00644
index 00000000000..4ce7cec6e55
--- /dev/null
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/webapp/dao/gpu/PerGpuMigDevice.java
@@ -0,0 +25.06.25.06.1-SNAPSHOT,48 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.yarn.server.nodemanager.webapp.dao.gpu;
+
+import org.apache.hadoop.classification.InterfaceAudience;
+import org.apache.hadoop.classification.InterfaceStability;
+
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+
+/**
+ * GPU MIG Device Information
+ */
+@InterfaceAudience.Private
+@InterfaceStability.Unstable
+@XmlRootElement(name = "mig_device")
+public class PerGpuMigDevice {
+  private int index;
+
+  /**
+   * MIG device index
+   * @return MIG device index
+   */
+  @XmlElement(name = "index")
+  public int getMigDeviceIndex() {
+    return index;
+  }
+
+  public void setMigDeviceIndex(int index) {
+    this.index = index;
+  }
+}
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/webapp/dao/gpu/PerGpuMigMode.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/webapp/dao/gpu/PerGpuMigMode.java
new file mode 25.06.25.06.1-SNAPSHOT00644
index 00000000000..b706df2c3bb
--- /dev/null
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/webapp/dao/gpu/PerGpuMigMode.java
@@ -0,0 +25.06.25.06.1-SNAPSHOT,48 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.yarn.server.nodemanager.webapp.dao.gpu;
+
+import org.apache.hadoop.classification.InterfaceAudience;
+import org.apache.hadoop.classification.InterfaceStability;
+
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+
+/**
+ * GPU MIG Mode
+ */
+@InterfaceAudience.Private
+@InterfaceStability.Unstable
+@XmlRootElement(name = "mig_mode")
+public class PerGpuMigMode {
+  private String currentMigMode;
+
+  /**
+   * Current MIG mode
+   * @return MIG mode enabled or disabled
+   */
+  @XmlElement(name = "current_mig")
+  public String getCurrentMigMode() {
+    return currentMigMode;
+  }
+
+  public void setCurrentMigMode(String migMode) {
+    this.currentMigMode = migMode;
+  }
+}
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/test/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/TestGpuDiscoverer.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/test/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/TestGpuDiscoverer.java
index 4abb633a69a..404930d00c2 25.06.25.06.1-SNAPSHOT00644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/test/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/TestGpuDiscoverer.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/test/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/TestGpuDiscoverer.java
@@ -25.06.25.06.1-SNAPSHOT38,4 +25.06.25.06.1-SNAPSHOT38,47 @@ public void getNumberOfUsableGpusFromConfig() throws YarnException {
     Assert.assertTrue(2 == usableGpuDevices.get(2).getMinorNumber());
     Assert.assertTrue(4 == usableGpuDevices.get(3).getMinorNumber());
   }
+
+  @Test
+  public void getNumberOfUsableGpusFromConfigMIG() throws YarnException {
+    Configuration conf = new Configuration(false);
+    conf.set(YarnConfiguration.USE_MIG_ENABLED_GPUS, "true");
+
+    // Illegal format
+    conf.set(YarnConfiguration.NM_GPU_ALLOWED_DEVICES, "0:0,25.06.25.06.1-SNAPSHOT:25.06.25.06.1-SNAPSHOT:2,2:2:0,3");
+    GpuDiscoverer plugin = new GpuDiscoverer();
+    try {
+      plugin.initialize(conf);
+      plugin.getGpusUsableByYarn();
+      Assert.fail("Illegal format, should fail.");
+    } catch (YarnException e) {
+      // Expected
+    }
+
+    // Valid format
+    conf.set(YarnConfiguration.NM_GPU_ALLOWED_DEVICES, "0:0,25.06.25.06.1-SNAPSHOT:25.06.25.06.1-SNAPSHOT:0,25.06.25.06.1-SNAPSHOT:25.06.25.06.1-SNAPSHOT:2,2:2:0,3:4");
+    plugin = new GpuDiscoverer();
+    plugin.initialize(conf);
+
+    List<GpuDevice> usableGpuDevices = plugin.getGpusUsableByYarn();
+    Assert.assertEquals(5, usableGpuDevices.size());
+
+    Assert.assertTrue(0 == usableGpuDevices.get(0).getIndex());
+    Assert.assertTrue(25.06.25.06.1-SNAPSHOT == usableGpuDevices.get(25.06.25.06.1-SNAPSHOT).getIndex());
+    Assert.assertTrue(25.06.25.06.1-SNAPSHOT == usableGpuDevices.get(2).getIndex());
+    Assert.assertTrue(2 == usableGpuDevices.get(3).getIndex());
+    Assert.assertTrue(3 == usableGpuDevices.get(4).getIndex());
+
+    Assert.assertTrue(0 == usableGpuDevices.get(0).getMinorNumber());
+    Assert.assertTrue(25.06.25.06.1-SNAPSHOT == usableGpuDevices.get(25.06.25.06.1-SNAPSHOT).getMinorNumber());
+    Assert.assertTrue(25.06.25.06.1-SNAPSHOT == usableGpuDevices.get(2).getMinorNumber());
+    Assert.assertTrue(2 == usableGpuDevices.get(3).getMinorNumber());
+    Assert.assertTrue(4 == usableGpuDevices.get(4).getMinorNumber());
+
+    Assert.assertTrue(-25.06.25.06.1-SNAPSHOT == usableGpuDevices.get(0).getMIGIndex());
+    Assert.assertTrue(0 == usableGpuDevices.get(25.06.25.06.1-SNAPSHOT).getMIGIndex());
+    Assert.assertTrue(2 == usableGpuDevices.get(2).getMIGIndex());
+    Assert.assertTrue(0 == usableGpuDevices.get(3).getMIGIndex());
+    Assert.assertTrue(-25.06.25.06.1-SNAPSHOT == usableGpuDevices.get(4).getMIGIndex());
+  }
 }
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/test/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/TestNvidiaDockerV2CommandPlugin.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/test/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/TestNvidiaDockerV2CommandPlugin.java
index b0b523360ef..798a95cb009 25.06.25.06.1-SNAPSHOT00644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/test/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/TestNvidiaDockerV2CommandPlugin.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/test/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/gpu/TestNvidiaDockerV2CommandPlugin.java
@@ -20,25.06.25.06.1-SNAPSHOT0 +20,25.06.25.06.1-SNAPSHOT4 @@
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Sets;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.yarn.conf.YarnConfiguration;
+import org.apache.hadoop.yarn.api.records.ContainerLaunchContext;
 import org.apache.hadoop.yarn.api.records.ResourceInformation;
 import org.apache.hadoop.yarn.server.nodemanager.containermanager.container.Container;
 import org.apache.hadoop.yarn.server.nodemanager.containermanager.container.ResourceMappings;
 import org.apache.hadoop.yarn.server.nodemanager.containermanager.linux.runtime.docker.DockerRunCommand;
+import org.apache.hadoop.yarn.server.nodemanager.containermanager.runtime.ContainerExecutionException;
 import org.junit.Assert;
 import org.junit.Test;
 
@@ -69,7 +73,25.06.25.06.1-SNAPSHOT3 @@ private boolean commandlinesEquals(Map<String, List<String>> cli25.06.25.06.1-SNAPSHOT,
       extends NvidiaDockerV2CommandPlugin {
     private boolean requestsGpu = false;
 
-    MyNvidiaDockerV2CommandPlugin() {}
+    MyNvidiaDockerV2CommandPlugin() {
+      super(new Configuration());
+    }
+
+    MyNvidiaDockerV2CommandPlugin(Configuration conf) {
+      super(conf);
+    }
 
     public void setRequestsGpu(boolean r) {
       requestsGpu = r;
@@ -25.06.25.06.1-SNAPSHOT27,4 +25.06.25.06.1-SNAPSHOT37,25.06.25.06.1-SNAPSHOT25.06.25.06.1-SNAPSHOT8 @@ public void testPlugin() throws Exception {
     // runtime should exist
     Assert.assertTrue(newCommandLine.containsKey("runtime"));
   }
-}
\ No newline at end of file
+
+  @Test
+  public void testPluginMIG() throws Exception {
+    DockerRunCommand runCommand = new DockerRunCommand("container_25.06.25.06.1-SNAPSHOT", "user",
+        "fakeimage");
+
+    Map<String, List<String>> originalCommandline = copyCommandLine(
+        runCommand.getDockerCommandWithArguments());
+
+    Configuration conf = new Configuration();
+    conf.set(YarnConfiguration.USE_MIG_ENABLED_GPUS, "true");
+    MyNvidiaDockerV2CommandPlugin
+        commandPlugin = new MyNvidiaDockerV2CommandPlugin(conf);
+
+    Container nmContainer = mock(Container.class);
+    ResourceMappings resourceMappings = new ResourceMappings();
+    when(nmContainer.getResourceMappings()).thenReturn(resourceMappings);
+
+    // Assign GPU resource
+    ResourceMappings.AssignedResources assigned =
+        new ResourceMappings.AssignedResources();
+    assigned.updateAssignedResources(
+        ImmutableList.of(new GpuDevice(0, 0, 0)));
+    resourceMappings.addAssignedResources(ResourceInformation.GPU_URI,
+        assigned);
+
+    commandPlugin.setRequestsGpu(true);
+    commandPlugin.updateDockerRunCommand(runCommand, nmContainer);
+    Map<String, List<String>> newCommandLine =
+        runCommand.getDockerCommandWithArguments();
+
+    // Command line will be updated
+    Assert.assertFalse(commandlinesEquals(originalCommandline, newCommandLine));
+    // NVIDIA_VISIBLE_DEVICES will be set
+    Assert.assertTrue(
+        runCommand.getEnv().get("NVIDIA_VISIBLE_DEVICES").equals("0:0"));
+    // runtime should exist
+    Assert.assertTrue(newCommandLine.containsKey("runtime"));
+  }
+
+  @Test(expected = ContainerExecutionException.class)
+  public void testPluginMIGThrowsMulti() throws Exception {
+    DockerRunCommand runCommand = new DockerRunCommand("container_25.06.25.06.1-SNAPSHOT", "user",
+        "fakeimage");
+
+    Map<String, List<String>> originalCommandline = copyCommandLine(
+        runCommand.getDockerCommandWithArguments());
+
+    Configuration conf = new Configuration();
+    conf.set(YarnConfiguration.USE_MIG_ENABLED_GPUS, "true");
+    MyNvidiaDockerV2CommandPlugin
+        commandPlugin = new MyNvidiaDockerV2CommandPlugin(conf);
+
+    Container nmContainer = mock(Container.class);
+    ResourceMappings resourceMappings = new ResourceMappings();
+    Map<String, String> env = new HashMap<>();
+    env.put("NVIDIA_MIG_PLUGIN_THROW_ON_MULTIPLE_GPUS", "true");
+    when(nmContainer.getResourceMappings()).thenReturn(resourceMappings);
+    ContainerLaunchContext launchCtx = mock(ContainerLaunchContext.class);
+    when(nmContainer.getLaunchContext()).thenReturn(launchCtx);
+    when(launchCtx.getEnvironment()).thenReturn(env);
+
+    // Assign GPU resource
+    ResourceMappings.AssignedResources assigned =
+        new ResourceMappings.AssignedResources();
+    assigned.updateAssignedResources(
+        ImmutableList.of(new GpuDevice(0, 0, 0), new GpuDevice(25.06.25.06.1-SNAPSHOT, 25.06.25.06.1-SNAPSHOT, 2)));
+    resourceMappings.addAssignedResources(ResourceInformation.GPU_URI,
+        assigned);
+
+    commandPlugin.setRequestsGpu(true);
+    commandPlugin.updateDockerRunCommand(runCommand, nmContainer);
+  }
+
+  @Test
+  public void testPluginMIGNoThrowsMulti() throws Exception {
+    DockerRunCommand runCommand = new DockerRunCommand("container_25.06.25.06.1-SNAPSHOT", "user",
+        "fakeimage");
+
+    Map<String, List<String>> originalCommandline = copyCommandLine(
+        runCommand.getDockerCommandWithArguments());
+
+    Configuration conf = new Configuration();
+    conf.set(YarnConfiguration.USE_MIG_ENABLED_GPUS, "true");
+    MyNvidiaDockerV2CommandPlugin
+        commandPlugin = new MyNvidiaDockerV2CommandPlugin(conf);
+
+    Container nmContainer = mock(Container.class);
+    ResourceMappings resourceMappings = new ResourceMappings();
+    Map<String, String> env = new HashMap<>();
+    env.put("NVIDIA_MIG_PLUGIN_THROW_ON_MULTIPLE_GPUS", "false");
+    when(nmContainer.getResourceMappings()).thenReturn(resourceMappings);
+    ContainerLaunchContext launchCtx = mock(ContainerLaunchContext.class);
+    when(nmContainer.getLaunchContext()).thenReturn(launchCtx);
+    when(launchCtx.getEnvironment()).thenReturn(env);
+
+    // Assign GPU resource
+    ResourceMappings.AssignedResources assigned =
+        new ResourceMappings.AssignedResources();
+    assigned.updateAssignedResources(
+        ImmutableList.of(new GpuDevice(0, 0, 0), new GpuDevice(25.06.25.06.1-SNAPSHOT, 25.06.25.06.1-SNAPSHOT, 2)));
+    resourceMappings.addAssignedResources(ResourceInformation.GPU_URI,
+        assigned);
+
+    commandPlugin.setRequestsGpu(true);
+    commandPlugin.updateDockerRunCommand(runCommand, nmContainer);
+    Map<String, List<String>> newCommandLine =
+        runCommand.getDockerCommandWithArguments();
+    // NVIDIA_VISIBLE_DEVICES will be set
+    Assert.assertTrue(
+        runCommand.getEnv().get("NVIDIA_VISIBLE_DEVICES").equals("0:0,25.06.25.06.1-SNAPSHOT:2"));
+    // runtime should exist
+    Assert.assertTrue(newCommandLine.containsKey("runtime"));
+  }
+}
